### PR TITLE
Feature/136892 136894 queue system authorization improvements

### PR DESCRIPTION
**Pull request by jesseeproductions**
_Ref:_ [C#136892](https://central.tri.be/issues/136892) and _Ref:_ [C#136894](https://central.tri.be/issues/136894) 

-Change to a Queue system instead of a Async
-Utilize Queue System for Custom Property Creation and Update
-Check for SSL
-Line Style Button
-Add Disconnect Button
-Add i Timeline Event Types for each Type and use Queue System to Create
-Check for existing Timeline Event Types and utilize ID instead of creating a new one
-Rename Contact_Property to  Contact_Properties to prevent autoloading conflicts with  Contact_Property_Group

### CODE REVIEW

File: src/Tribe/API/Timeline.php [hunk 1]
```diff
@@ -11,7 +12,31 @@
...
 class Timeline {
 
-	protected $properties = [];
+	/**
+	 * @var array
+	 */
+	protected $timeline_event_types = [
+		'eventRegistration'     => [
+			'site_option'    => 'timeline_event_registration_id',
+			'headerTemplate' => 'Ticket purchase occurred at {{#formatDate timestamp}}{{/formatDate}} from the Event Tickets HubSpot Integration app',
```
Comment (sc0ttkclark) [inline src/Tribe/API/Timeline.php]: Should these be defined in a method and use `__()`?

File: src/Tribe/API/Timeline.php [hunk 2]
```diff
@@ -21,46 +46,90 @@ class Timeline {
...
+		$queue = new Delivery_Queue();
+		$queue->push_to_queue( $hubspot_data );
+		$queue->save();
+		$queue->dispatch();
 
 	}
 
 	/**
-	 * Create Timeline Event Type
+	 * Create all Timeline Event Types
```
Comment (sc0ttkclark) [inline src/Tribe/API/Timeline.php]:
```suggestion
	 * Create all Timeline Event Types.
```

File: src/Tribe/API/Timeline.php [hunk 3]
```diff
@@ -21,46 +46,90 @@ class Timeline {
...
 		$hubspot_api = tribe( 'tickets.hubspot.api' );
 
 		if ( ! $access_token = $hubspot_api->is_ready() ) {
 			return false;
 		}
 
-		$client = $hubspot_api->client;
-		$appId  = isset( $options['app_id'] ) ? $options['app_id'] : '';
+		$client          = $hubspot_api->client;
+		$appId           = isset( $options['app_id'] ) ? (int) $options['app_id'] : '';
```
Comment (sc0ttkclark) [inline src/Tribe/API/Timeline.php]:
```suggestion
		$app_id           = isset( $options['app_id'] ) ? (int) $options['app_id'] : '';
```

File: src/Tribe/API/Timeline.php [hunk 4]
```diff
@@ -21,46 +46,90 @@ class Timeline {
...
+			$matching_event_types = wp_filter_object_list( $timeline_events, [ 'name' => $name ], 'and' );
+			$matching_event_type  = reset( $matching_event_types );
 
-		//todo setup all 3 types in another class and generate each one here
-		$name           = 'eventRegistration';
-		$headerTemplate = 'Ticket purchase occurred at {{#formatDate timestamp}}{{/formatDate}} from the Event Tickets HubSpot Integration app';
-		$detailTemplate = 'Purchased ticket for {{extraData.event.post_title}} (# {{extraData.event.ID}} )';
+			// If the match is empty then there is no event type and create it.
+			if ( empty( $matching_event_type->applicationId ) ) {
+				$this->create_event_type( $hubspot_options, $access_token, $client, $appId, $name, $event_type );
```
Comment (sc0ttkclark) [inline src/Tribe/API/Timeline.php]:
```suggestion
				$this->create_event_type( $hubspot_options, $access_token, $client, $app_id, $name, $event_type );
```

File: src/Tribe/API/Timeline.php [hunk 5]
```diff
@@ -21,46 +46,90 @@ class Timeline {
...
+				continue;
+			}
+
+			// If the match and the app id does not match then create it.
+			if ( $appId !== (int) $matching_event_type->applicationId ) {
```
Comment (sc0ttkclark) [inline src/Tribe/API/Timeline.php]:
```suggestion
			if ( $app_id !== (int) $matching_event_type->applicationId ) {
```

File: src/Tribe/API/Timeline.php [hunk 6]
```diff
@@ -21,46 +46,90 @@ class Timeline {
...
+				$this->create_event_type( $hubspot_options, $access_token, $client, $appId, $name, $event_type );
```
Comment (sc0ttkclark) [inline src/Tribe/API/Timeline.php]:
```suggestion
				$this->create_event_type( $hubspot_options, $access_token, $client, $app_id, $name, $event_type );
```

File: src/Tribe/API/Timeline.php [hunk 7]
```diff
@@ -21,46 +46,90 @@ class Timeline {
...
+	}
+
+	/**
+	 * Create Timeline Event Type
+	 *
+	 * @since 1.0
+	 *
+	 * @return bool
+	 */
+	public function create_event_type( $hubspot_options, $access_token, $client, $appId, $name, $event_type ) {
```
Comment (sc0ttkclark) [inline src/Tribe/API/Timeline.php]:
```suggestion
	public function create_event_type( $hubspot_options, $access_token, $client, $app_id, $name, $event_type ) {
```

File: src/Tribe/API/Timeline.php [hunk 8]
```diff
@@ -21,46 +46,90 @@ class Timeline {
...
 
 		try {
 			$hubspot  = Factory::createWithToken( $access_token, $client );
-			$response = $hubspot->Timeline()->createEventType( $appId, $name, $headerTemplate, $detailTemplate );
+			$response = $hubspot->Timeline()->createEventType( $appId, $name, $event_type['headerTemplate'], $event_type['detailTemplate'] );
```
Comment (sc0ttkclark) [inline src/Tribe/API/Timeline.php]:
```suggestion
			$response = $hubspot->Timeline()->createEventType( $app_id, $name, $event_type['headerTemplate'], $event_type['detailTemplate'] );
```

File: src/Tribe/API/Timeline.php [hunk 9]
```diff
@@ -21,46 +46,90 @@ class Timeline {
...
+				$hubspot_options->update_option( $event_type['site_option'], $matching_event_type->id );
+				continue;
+			}
+
+		}
+	}
+
+	/**
+	 * Create Timeline Event Type
+	 *
```
Comment (sc0ttkclark) [inline src/Tribe/API/Timeline.php]: Needs phpdoc for method arguments

File: src/Tribe/Process/Delivery_Queue.php [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+
+	/**
+	 * Returns the queue process action name.
+	 *
+	 * @since 1.0
+	 *
+	 * @return string
+	 */
+	public static function action() {
+		return 'hubspot_delivery_Queue';
```
Comment (sc0ttkclark) [inline src/Tribe/Process/Delivery_Queue.php]: Intentionally capitalized Queue?

File: src/Tribe/API/Timeline.php (hunk 1 above)
Comment (jesseeproductions) [reply src/Tribe/API/Timeline.php]: They are the templates that are saved in HubSpot on the creation of each event type. While they could be translated there is no coding to update them in HubSpot if translated in the future or a way then for someone on a multi language site to choose which one they would like. These templates once saved in HubSpot can be edited in the app in HubSpot so if someone would like to change what it says they have the opportunity to do it there.

File: src/Tribe/Process/Delivery_Queue.php (hunk 10 above)
Comment (jesseeproductions) [reply src/Tribe/Process/Delivery_Queue.php]: nope, PHPstorm refactor issue.

**Pull request merged**